init.lua

```diff
@@ -169,61 +167,65 @@ function ranks.update_nametag(player)
 			prefix = ""
 		end
 
-		player:set_nametag_attributes({
-			text = prefix..name,
-		})
+		if player then
+			player:set_nametag_attributes({
+				text = prefix..name,
+			})
+		end
 
 		return true
 	end
 end
 
 -- [function] Set player rank
-function ranks.set_rank(player, rank)
-	if type(player) == "string" then
-		player = minetest.get_player_by_name(player)
+function ranks.set_rank(name, rank)
+	if type(name) ~= "string" then
+		name = name:get_player_name()
 	end
 
-	if registered[rank] then
-		-- Set attribute
-		player:set_attribute("ranks:rank", rank)
+	if registered[rank] and minetest.player_exists(name) then
+		storage:set_string(name, rank)
+
 		-- Update nametag
-		ranks.update_nametag(player)
+		ranks.update_nametag(name)
 		-- Update privileges
-		ranks.update_privs(player)
+		ranks.update_privs(name)
 
 		return true
 	end
 end
 
 -- [function] Remove rank from player
-function ranks.remove_rank(player)
-	if type(player) == "string" then
-		player = minetest.get_player_by_name(player)
+function ranks.remove_rank(name)
+	if type(name) ~= "string" then
+		player = name:get_player_name()
+	else
+		player = minetest.get_player_by_name(name)
 	end
 
-	local rank = ranks.get_rank(player)
-	if rank then
-		local name = player:get_player_name()
-
-		-- Clear attribute
-		player:set_attribute("ranks:rank", nil)
-		-- Update nametag
-		player:set_nametag_attributes({
-			text = name,
-			color = "#ffffff",
-		})
-		-- Update privileges
-		local basic_privs =
-			minetest.string_to_privs(minetest.settings:get("basic_privs") or "interact,shout")
-		minetest.set_player_privs(name, basic_privs)
+	local rank = ranks.get_rank(name)
+	if rank ~= nil then
+		storage:set_string(name, nil)
+
+		if player then
```

Comment:
...and `player` should then be defined on the line before this if statement begins (i.e. line 209)
```lua
local player = minetest.get_player_by_name(name)
```